core/src/broadcast_stage/slot_transmit_shreds_cache.rs

```diff
@@ -0,0 +1,675 @@
+                .unwrap_or(false)
+    }
+
+    pub fn to_transmit_shreds(&self) -> Vec<TransmitShreds> {
+        self.data_shred_batches
+            .iter()
+            .map(|data_shred_batch| (self.stakes.clone(), data_shred_batch.clone()))
+            .chain(
+                self.coding_shred_batches
+                    .iter()
+                    .map(|code_shred_batch| (self.stakes.clone(), code_shred_batch.clone())),
+            )
+            .collect()
+    }
+
+    pub fn last_data_shred(&self) -> Option<&Shred> {
+        self.data_shred_batches
+            .last()
+            .and_then(|last_shred_batch| last_shred_batch.last())
+    }
+
+    pub fn last_coding_shred(&self) -> Option<&Shred> {
+        self.coding_shred_batches
+            .last()
+            .and_then(|last_shred_batch| last_shred_batch.last())
+    }
+
+    // Updates with new shreds from blockstore and returns a `SlotCachedTransmitShreds`
+    // representing just the diff from the updates
+    pub fn update_from_blockstore(&mut self, blockstore: &Blockstore) -> Option<Self> {
+        if !self.contains_last_shreds() {
+            let last_data_shred_index = self
+                .last_data_shred()
+                .map(|shred| shred.index() + 1)
+                .unwrap_or(0);
+
+            let last_coding_shred_index = self
+                .last_coding_shred()
+                .map(|shred| shred.index() + 1)
+                .unwrap_or(0);
+
+            let (new_data_shreds, new_coding_shreds) = Self::get_new_shreds(
+                blockstore,
+                self.slot,
+                last_data_shred_index as u64,
+                last_coding_shred_index as u64,
+            );
+
+            let data_shred_batches = {
+                if !new_data_shreds.is_empty() {
+                    self.data_shred_batches.push(new_data_shreds.clone());
+                    vec![new_data_shreds]
+                } else {
+                    vec![]
+                }
+            };
+
+            let coding_shred_batches = {
+                if !new_coding_shreds.is_empty() {
+                    self.coding_shred_batches.push(new_coding_shreds.clone());
+                    vec![new_coding_shreds]
+                } else {
+                    vec![]
+                }
+            };
+
+            if !data_shred_batches.is_empty() || !coding_shred_batches.is_empty() {
+                Some(SlotCachedTransmitShreds {
+                    slot: self.slot,
+                    stakes: self.stakes.clone(),
+                    data_shred_batches,
+                    coding_shred_batches,
+                })
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+
+    fn get_new_shreds(
+        blockstore: &Blockstore,
+        slot: Slot,
+        data_start_index: u64,
+        coding_start_index: u64,
+    ) -> (Arc<Vec<Shred>>, Arc<Vec<Shred>>) {
+        let new_data_shreds = Arc::new(
```

Comment:
Why not have this thing give blockstore a vec to push onto. Then we don't have this batches thing, only need to copy shreds once instead of twice.